packages/lib/src/load_plugin_list.ts

```diff
@@ -0,0 +1,134 @@
+import fs from "fs-extra";
+import path from "path";
+import { logger } from "./logging";
+import * as libFileOps from "./file_ops";
+
+/**
+ * Searches for local plugins in the plugins/ and external_plugins/ directories
+ * and adds them to the plugin list if they contain a package.json file.
+ *
+ * @param pluginList - Map of plugin names to their file paths
+ * @param pluginListPath - Path to the JSON file storing the plugin list
+ * @returns Promise that resolves when plugin discovery is complete
+ */
+async function findLocalPlugins(pluginList: Map<string, string>, pluginListPath: string) {
+	// Check folders for plugins
+	const pluginFolders = [
+		"plugins",
+		"external_plugins",
+	];
+
+	let has_changed = 0;
+	for (const pluginFolder of pluginFolders) {
+		let pluginNames;
+		try {
+			pluginNames = await fs.readdir(pluginFolder);
+		} catch (err: any) {
+			if (err.code === "ENOENT") {
+				// Skip if folder doesn't exist
+				continue;
+			}
+			throw err;
+		}
+		for (const pluginName of pluginNames) {
+			if (pluginList.has(pluginName)) {
+				continue;
+			}
+			const pluginPath = path.resolve(pluginFolder, pluginName);
+			const stats = await fs.stat(pluginPath);
+			if (!stats.isDirectory()) {
+				continue;
+			}
+
+			// Check that the plugin has a package.json file
+			const packageJsonPath = path.resolve(pluginPath, "package.json");
+			if (!(await fs.exists(packageJsonPath))) {
+				continue;
+			}
```

Comment:
I hadn't considered mono repo support. We don't want to crawl too deep because it increases the risk of a false positive and hurts performance. Would 1 layer more be sufficient?

Reply:
Although 1 level would be sufficient for me where I have `external_plugins/repo/plugin_name`. I can see the possibility of someone wanting `external_plugins/repo/plugins/plugin_name` which would require 2 levels of search. I do not see any reason to support more than this. I also do not expect too much performance impact if we only explore deeper if the folder is not a plugin.

If we will need to check for the keyword anyway, to avoid adding a mono-repo workspace as a plugin, then we can also validate that the "clusterio-plugin" keyword still exists and remove the plugin if it doesn't. Would you accept removing the keyword from the package json to disable it during development? Alternatively could you move your disabled folder to be outside of external_plugins? ie rather than `clusterio/external_plugins/disabled` you have `clusterio/disabled_plugins` ?

Reply:
Moving the disabled folder also works. Not loading plugins without the keyword also works, but is definitely a breaking change.

Reply:
I agree with you on monorepo performance - as long as we ignore common folders like node_modules and .git and stop when we hit the first plugin it should be fine. I'll add recursion.